### ISSUE TITLE
ABI does not contain sufficient information to compute the correct function selector

### ISSUE THREAD

**Issue by gsalzer** [ethereum/solidity#7530]
## Description

The ABI does not contain sufficient information to compute the correct function selector for accessing the contract. More specifically, a `library` and a `contract` with the same entry points may have different 4 byte selectors, but nothing in the ABI allows to decide whether we are dealing with a contract or a libary.

A similar issue is raised in #7102, but there the conclusion is to put the information into the metadata, which does not the issue here. Possible solutions:

- Put a flag like `contractKind: contract` or `contractKind: library` into the ABI.
- For each function (and event?), add a field with the string that has to be hashed for the selector.
- Less preferably: for each function (and event?), add a field with the 4/32-byte selector.

## Environment

- Compiler version: 0.5.12
- Target EVM version (as per compiler settings): default
- Framework/IDE (e.g. Truffle or Remix): command line
- EVM execution environment / backend / blockchain client:
- Operating system: Linux

## Steps to Reproduce

The ABIs of the following two contracts are identical, but the selectors are different.
```
$ cat l.sol
pragma solidity ^0.5.0;
pragma experimental ABIEncoderV2;
library C {
  struct S {address a;}
  function F(S calldata) external pure returns (uint256) {return 0;}
}

$ solc --hashes --abi l.sol
[...]
Function signatures: 
9b21e02d: F(C.S)
Contract JSON ABI 
[{"constant":true,"inputs":[{"components":[{"internalType":"address","name":"a","type":"address"}],"internalType":"struct C.S","name":"","type":"tuple"}],"name":"F","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"payable":false,"stateMutability":"pure","type":"function"}]
```
```
$ cat c.sol
pragma solidity ^0.5.0;
pragma experimental ABIEncoderV2;
contract C {
  struct S {address a;}
  function F(S calldata) external pure returns (uint256) {return 0;}
}

$ solc --hashes --abi c.sol
[...]
Function signatures: 
a0b61132: F((address))
Contract JSON ABI 
[{"constant":true,"inputs":[{"components":[{"internalType":"address","name":"a","type":"address"}],"internalType":"struct C.S","name":"","type":"tuple"}],"name":"F","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"payable":false,"stateMutability":"pure","type":"function"}]
```

**Comment by chriseth** [comment]
This is a bug in the compiler, the function should not be part of the library's ABI, because its hash is computed differently. Usually, library functions are not supposed to be called directly (i.e. not from a contract), but since people are doing it, we allowed it for some functions.

**Comment by gsalzer** [comment]
My impression was that `external`/`public` functions of libraries should be callable by  contracts via `delegatecall`. Moreover, shouldn't libraries be usable from contracts that have been compiled by other tools than Solidity (e.g., by contracts in other languages)? In this case the ABI should contain all external functions, sufficient information for computing the correct selector, and information on the admissible types of calls.

**Comment by chriseth** [comment]
Indeed, but libraries are considered to be solidity-internal. They have always used features that were not part of the generic ABI (storage pointers, for instance).

Having said that, I think there is also another issue planned for 0.6.0 that should solve this problem.

**Comment by erak** [comment]
I think this is related to https://github.com/ethereum/solidity/issues/6420.

**Comment by chriseth** [comment]
Partly solved by #7719

**Comment by axic** [comment]
What is the status here? Do we want to start omitting such library functions from the ABI in 0.7.0?

cc @chriseth @ekpyron

**Comment by Derixtar54** [comment]
The ABI does not contain sufficient information to compute the correct function selector for accessing the contract. More specifically, a library and a contract with the same entry points may have different 4 byte selectors, but nothing in the ABI allows to decide whether we are dealing with a contract or a libary.

A similar issue is raised in https://github.com/ethereum/solidity/issues/7102, but there the conclusion is to put the information into the metadata, which does not the issue here. Possible solutions:

Put a flag like contractKind: contract or contractKind: library into the ABI.
For each function (and event?), add a field with the string that has to be hashed for the selector.
Less preferably: for each function (and event?), add a field with the 4/32-byte selector.
Environment
Compiler version: 0.5.12
Target EVM version (as per compiler settings): default
Framework/IDE (e.g. Truffle or Remix): command line
EVM execution environment / backend / blockchain client:
Operating system: Linux
Steps to Reproduce
The ABIs of the following two contracts are identical, but the selectors are different.

$ cat l.sol
pragma solidity ^0.5.0;
pragma experimental ABIEncoderV2;
library C {
  struct S {address a;}
  function F(S calldata) external pure returns (uint256) {return 0;}
}

$ solc --hashes --abi l.sol
[...]
Function signatures: 
9b21e02d: F(C.S)
Contract JSON ABI 
[{"constant":true,"inputs":[{"components":[{"internalType":"address","name":"a","type":"address"}],"internalType":"struct C.S","name":"","type":"tuple"}],"name":"F","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"payable":false,"stateMutability":"pure","type":"function"}]
$ cat c.sol
pragma solidity ^0.5.0;
pragma experimental ABIEncoderV2;
contract C {
  struct S {address a;}
  function F(S calldata) external pure returns (uint256) {return 0;}
}

$ solc --hashes --abi c.sol
[...]
Function signatures: 
a0b61132: F((address))
Contract JSON ABI 
[{"constant":true,"inputs":[{"components":[{"internalType":"address","name":"a","type":"address"}],"internalType":"struct C
